python/Ganga/GPIDev/Base/Objects.py

```diff
@@ -74,6 +73,7 @@ class Node(object):
     thread-safe usage.
     """
     __metaclass__ = abc.ABCMeta
+    __slots__ = ('_index_cache_dict', '_parent', '_read_lock', '_write_lock', '_dirty', '_registry', '_data_dict')
```

Comment:
> It also prevents us from ever dynamically adding attributes to objects of this type.

Isn't this a good thing? Certainly this and the memory performance improvements that come with not having to define a `__dict__` for each instantiated object are the reasons that I used them in the stuff I've been doing.

My argument would be that only the schema attributes should be needed by the user for a given object and that they shouldn't be adding extras. Obviously as a developer if we want an attribute added to a derived class then we can add it to the `__slots__` for that class. I don't really like the idea of dynamic attributes as it's hard to say at any given time what an object contains.